### PR TITLE
Paypal integration

### PR DESCRIPTION
**Pull request by rupertbates**
This is stage one of the work to integrate PayPal payments into membership-frontend it depends on [this PR](https://github.com/guardian/membership-common/pull/373) in membership-common.

At the moment there are no UI changes this PR simply introduces a new form field 'payPalBaid' which may be present in the payment form submitted to Joiner.enterPaidDetails, the original token field which represented a token in the Stripe system is now an Option and has been renamed to 'stripeToken'.

Depending on which of these tokens is present the code will create a user with either a PayPal payment method or a Stripe payment method.

@rtyley @paulbrown1982 @johnduffell could one or more of you take a look for me please?
cc/ @Ap0c @svillafe

### CODE REVIEW

**Comment by Ap0c** [comment]
For anyone wondering about the failing tests, that's happening because these changes rely on a version of `membership-common` that hasn't been published yet, so sbt can't resolve the dependency. They *should* pass ok once the PR that @rupertbates mentioned above gets merged.

File: frontend/app/services/MemberService.scala [hunk 1]
```diff
@@ -110,73 +111,85 @@ class MemberService(identityService: IdentityService,
...
-          stripeCustomer  <- createStripeCustomer(paid)
-          idUser          <- getIdentityUserDetails()
-          sfContact       <- createSalesforceContact(idUser)
-          zuoraSubName    <- createPaidZuoraSubscription(sfContact, stripeCustomer, paid, idUser.primaryEmailAddress)
-          _               <- updateSalesforceContactWithMembership(Some(stripeCustomer))  // FIXME: This should go!
-          _               <- updateIdentity()
+          idUser <- getIdentityUserDetails()
+          sfContact <- createSalesforceContact(idUser)
+          zuoraSubName <- createPaidZuoraSubscription(sfContact, paid, idUser.primaryEmailAddress, None)
+          _ <- updateSalesforceContactWithMembership(None) // FIXME: This should go!
```
Comment (paulbrown1982) [inline frontend/app/services/MemberService.scala]: FYI, the FIXME is covered here: https://trello.com/c/6EXHhV6x/99-remove-salesforce-call-to-update-membership-tier-in-contact-object

File: frontend/app/services/MemberService.scala [hunk 2]
```diff
@@ -390,46 +405,53 @@ class MemberService(identityService: IdentityService,
...
 
+  private def createAccount(contactId: ContactId, currency: Currency, paymentForm: PaymentForm) =
+    paymentForm match {
+      case PaymentForm(_, Some(stripeToken), _) =>
+        Account.stripe(contactId, currency, true)
+      case PaymentForm(_, _, Some(payPalBaid)) =>
+        Account.payPal(contactId, currency, true)
+    }
+
+  private def createPaymentMethod(contactId: ContactId, email: String, paymentForm: PaymentForm, stripeCustomer: Option[Customer]) =
```
Comment (paulbrown1982) [inline frontend/app/services/MemberService.scala]: 👍 Nice that this is now in its own method.

File: frontend/app/services/MemberService.scala [hunk 3]
```diff
@@ -390,46 +405,53 @@ class MemberService(identityService: IdentityService,
...
-        ).some,
+
+      Subscribe(account = createAccount(contactId, currency, joinData.payment),
+        paymentMethod = createPaymentMethod(contactId, email, joinData.payment, stripeCustomer),
         address = joinData.zuoraAccountAddress,
         email = email,
         ratePlans = NonEmptyList(plan),
-        name = nameData,
-        ipAddress = ipAddress.map(_.getHostAddress),
-        ipCountry = ipCountry
```
Comment (Ap0c) [inline frontend/app/services/MemberService.scala]: I think the changes made to drop `ipAddress` and `ipCountry` may conflict with the changes made to this file in #1401. @paulbrown1982, is that a possibility?

File: project/Dependencies.scala [hunk 4]
```diff
@@ -9,7 +9,7 @@ object Dependencies {
   val sentryRavenLogback = "com.getsentry.raven" % "raven-logback" % "7.2.3"
   val scalaUri = "com.netaporter" %% "scala-uri" % "0.4.6"
   val memsubCommonPlayAuth = "com.gu" %% "memsub-common-play-auth" % "0.7"
-  val membershipCommon = "com.gu" %% "membership-common" % "0.317"
+  val membershipCommon = "com.gu" %% "membership-common" % "0.320-SNAPSHOT"
```
Comment (Ap0c) [inline project/Dependencies.scala]: Am I right in thinking that the changes to common will now be published as `0.319`? Except if @pvighi merges and releases [his changes](https://github.com/guardian/membership-common/pull/370) first, and then it should be `0.320`?

File: frontend/app/services/MemberService.scala (hunk 3 above)
Comment (paulbrown1982) [inline frontend/app/services/MemberService.scala]: @Ap0c @rupertbates We should definitely keep ipCountry carrying through, but not ipAddress.

File: project/Dependencies.scala (hunk 4 above)
Comment (johnduffell) [inline project/Dependencies.scala]: I was a bit confused! and yes, just fill in this number when you know it, then push that change.

File: frontend/app/services/MemberService.scala [hunk 5]
```diff
@@ -110,73 +111,85 @@ class MemberService(identityService: IdentityService,
...
 
     def updateSalesforceContactWithMembership(stripeCustomer: Option[Customer]): Future[ContactId] =
       salesforceService.updateMemberStatus(user, tier, stripeCustomer).andThen { case Failure(e) =>
-        logger.error(s"Could not update Salesforce contact with membership status for user ${user.id}", e)}
+        logger.error(s"Could not update Salesforce contact with membership status for user ${user.id}", e)
+      }
 
     formData match {
-      case paid: PaidMemberJoinForm => //
+      case paid@PaidMemberJoinForm(_, _, PaymentForm(_, _, Some(_)), _, _, _, _, _, _, _, _, _) => //Paid member with PayPal token
```
Comment (rupertbates) [inline frontend/app/services/MemberService.scala]: _* doesn't work in a match against a case class

Comment (Ap0c) [inline frontend/app/services/MemberService.scala]: Yup, you're right, seems to only work as varargs in sequences. Sorry, ignore this comment then 😬.

**Comment by Ap0c** [comment]
Just signed up a test supporter in PROD, worked ok!

**Pull request merged**